### PR TITLE
Add Optional ECN MetaData to Addressed Envelope

### PR DESCRIPTION
**Pull request by PeterAdams-A**
Add Optional ECN MetaData to Addressed Envelope

### Motivation:

This is a very small step along the path to https://github.com/apple/swift-nio/issues/1472

### Modifications:

Add an optional metadata field to addressed envelope.  Currently the only possible data is ECN information.  This is not currently populated anywhere.

### Result:

Metadata in place for future work.

### CODE REVIEW

File: Sources/NIO/AddressedEnvelope.swift [hunk 1]
```diff
@@ -20,15 +20,35 @@
...
+    /// Any MetaData associated with this `AddressedEnvelope`
+    public var metaData : MetaData? = nil
 
     public init(remoteAddress: SocketAddress, data: DataType) {
         self.remoteAddress = remoteAddress
         self.data = data
     }
+    
+    /// Any MetaData associated with an AddressedEnvelope
+    public struct MetaData {
```
Comment (Lukasa) [inline Sources/NIO/AddressedEnvelope.swift:32]: `Metadata` is one word, and so should be spelled...well, the way I just spelled it.

File: Sources/NIO/AddressedEnvelope.swift [hunk 2]
```diff
@@ -20,15 +20,35 @@
...
 }
 
 extension AddressedEnvelope: CustomStringConvertible {
     public var description: String {
         return "AddressedEnvelope { remoteAddress: \(self.remoteAddress), data: \(self.data) }"
     }
 }
+
+/// Possible ECN States
+public enum NIOEcnState {
```
Comment (Lukasa) [inline Sources/NIO/AddressedEnvelope.swift:45]: As "ECN" is an acronym this should read `NIOECNState`.

Comment (Lukasa) [reply Sources/NIO/AddressedEnvelope.swift:45]: Can this be conformed to `Hashable`?

File: Sources/NIO/AddressedEnvelope.swift [hunk 3]
```diff
@@ -20,15 +20,35 @@
 public struct AddressedEnvelope<DataType> {
     public var remoteAddress: SocketAddress
     public var data: DataType
+    /// Any MetaData associated with this `AddressedEnvelope`
+    public var metaData : MetaData? = nil
```
Comment (Lukasa) [inline Sources/NIO/AddressedEnvelope.swift:24]: Similarly, this should be `metadata`.

Comment (PeterAdams-A) [reply Sources/NIO/AddressedEnvelope.swift]: done.

File: Sources/NIO/AddressedEnvelope.swift (hunk 2 above)
Comment (PeterAdams-A) [reply Sources/NIO/AddressedEnvelope.swift]: done.

File: Sources/NIO/AddressedEnvelope.swift (hunk 1 above)
Comment (PeterAdams-A) [reply Sources/NIO/AddressedEnvelope.swift]: done.

File: Sources/NIO/AddressedEnvelope.swift [hunk 4]
```diff
@@ -20,15 +20,35 @@
 public struct AddressedEnvelope<DataType> {
     public var remoteAddress: SocketAddress
     public var data: DataType
+    /// Any metadata associated with this `AddressedEnvelope`
+    public var metadata : Metadata? = nil
```
Comment (Lukasa) [inline Sources/NIO/AddressedEnvelope.swift:24]: Mild nit

```suggestion
    public var metadata: Metadata? = nil
```

File: Sources/NIO/AddressedEnvelope.swift [hunk 5]
```diff
@@ -20,15 +20,35 @@
...
 
     public init(remoteAddress: SocketAddress, data: DataType) {
         self.remoteAddress = remoteAddress
         self.data = data
     }
+    
+    /// Any metadata associated with an `AddressedEnvelope`
+    public struct Metadata {
+        /// Details of any congestion state.
+        public var ecnState : NIOECNState
```
Comment (Lukasa) [inline Sources/NIO/AddressedEnvelope.swift:34]: Mild nit

```suggestion
        public var ecnState: NIOECNState
```

File: Sources/NIO/AddressedEnvelope.swift (hunk 4 above)
Comment (glbrntt) [inline Sources/NIO/AddressedEnvelope.swift:24]:
```suggestion
    public var metadata: Metadata? = nil
```

File: Sources/NIO/AddressedEnvelope.swift (hunk 5 above)
Comment (glbrntt) [inline Sources/NIO/AddressedEnvelope.swift:34]:
```suggestion
        public var ecnState: NIOECNState
```

File: Sources/NIO/AddressedEnvelope.swift [hunk 6]
```diff
@@ -20,15 +20,35 @@
...
+/// Possible ECN States
+public enum NIOECNState: Hashable {
+    /// Non-ECN Capable Transport.
+    case nonECT
+    /// ECN Capable Transport (flag 0).
+    case ect0
+    /// ECN Capable Transport (flag 1).
+    case ect1
+    /// Congestion Experienced.
+    case ce
```
Comment (glbrntt) [inline Sources/NIO/AddressedEnvelope.swift:53]: nit: `congestionEncountered`/`congestionExperienced`

Comment (PeterAdams-A) [reply Sources/NIO/AddressedEnvelope.swift:53]: I was heavily influenced by
https://developer.apple.com/documentation/network/nwprotocolip/ecn which has experienced.
Wikipedia https://en.wikipedia.org/wiki/Explicit_Congestion_Notification goes for encountered.
RFC https://tools.ietf.org/html/rfc3168 goes for experienced.

I think experienced wins 2:1 :-)

Comment (Lukasa) [reply Sources/NIO/AddressedEnvelope.swift:53]: I think George is proposing changing the name of the enum case. ;)

Comment (glbrntt) [reply Sources/NIO/AddressedEnvelope.swift:53]: Correct. We should spell it out either way to gain some clarity here.

Comment (PeterAdams-A) [reply Sources/NIO/AddressedEnvelope.swift:53]: Ahh - I should probably spell them all out then.

Comment (PeterAdams-A) [reply Sources/NIO/AddressedEnvelope.swift]: I have expanded the naming in this area.

File: Sources/NIO/AddressedEnvelope.swift [hunk 7]
```diff
@@ -20,15 +20,35 @@
...
+    /// Any metadata associated with this `AddressedEnvelope`
+    public var metadata: Metadata? = nil
 
     public init(remoteAddress: SocketAddress, data: DataType) {
         self.remoteAddress = remoteAddress
         self.data = data
     }
+    
+    /// Any metadata associated with an `AddressedEnvelope`
+    public struct Metadata {
```
Comment (Lukasa) [inline Sources/NIO/AddressedEnvelope.swift:32]: Let's make this struct `Hashable` too, while we're here.

Comment (PeterAdams-A) [reply Sources/NIO/AddressedEnvelope.swift]: Done.

**Pull request merged**